File: src/chapters/minimization/content/code/simplex.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// Nelder-Mead downhill simplex minimization.
func nelderMead(f func([]float64) float64, x0 []float64, step, tol float64, maxIter int) []float64 {
	n := len(x0)
	pts := [][]float64{append([]float64{}, x0...)}
	for i := 0; i < n; i++ {
		p := append([]float64{}, x0...)
		p[i] += step
		pts = append(pts, p)
	}
	fv := make([]float64, n+1)
	for i := range pts {
		fv[i] = f(pts[i])
	}
	for it := 0; it < maxIter; it++ {
		ord := make([]int, n+1)
		for i := range ord {
			ord[i] = i
		}
		sort.Slice(ord, func(a, b int) bool { return fv[ord[a]] < fv[ord[b]] })
		np := make([][]float64, n+1)
		nf := make([]float64, n+1)
		for i, o := range ord {
			np[i] = pts[o]
			nf[i] = fv[o]
		}
		pts, fv = np, nf
		if math.Abs(fv[n]-fv[0]) < tol {
			break
		}
		c := make([]float64, n) // centroid of best n points
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				c[j] += pts[i][j] / float64(n)
			}
		}
		xr := make([]float64, n) // reflect
		for j := range xr {
			xr[j] = c[j] + (c[j] - pts[n][j])
		}
		fr := f(xr)
		if fr < fv[0] {
			xe := make([]float64, n) // expand
			for j := range xe {
				xe[j] = c[j] + 2*(c[j]-pts[n][j])
			}
			if fe := f(xe); fe < fr {
				pts[n], fv[n] = xe, fe
			} else {
				pts[n], fv[n] = xr, fr
			}
		} else if fr < fv[n-1] {
			pts[n], fv[n] = xr, fr
		} else {
			xc := make([]float64, n) // contract
			for j := range xc {
				xc[j] = c[j] + 0.5*(pts[n][j]-c[j])
			}
			if fc := f(xc); fc < fv[n] {
				pts[n], fv[n] = xc, fc
			} else { // shrink toward best
				for i := 1; i <= n; i++ {
					for j := 0; j < n; j++ {
						pts[i][j] = pts[0][j] + 0.5*(pts[i][j]-pts[0][j])
					}
					fv[i] = f(pts[i])
				}
			}
		}
	}
	return pts[0]
}

func main() {
	f := func(v []float64) float64 { return (v[0]-1)*(v[0]-1) + (v[1]-2)*(v[1]-2) }
	fmt.Println(nelderMead(f, []float64{0, 0}, 0.5, 1e-10, 400)) // -> [1 2]
}

File: src/chapters/minimization/content/code/simplex.js
// Nelder-Mead downhill simplex minimization.
function nelderMead(f, x0, step = 0.5, tol = 1e-10, maxIter = 400) {
  const n = x0.length;
  let pts = [[...x0]];
  for (let i = 0; i < n; i++) {
    const p = [...x0]; p[i] += step; pts.push(p);
  }
  let fv = pts.map((p) => f(p));
  for (let it = 0; it < maxIter; it++) {
    const ord = [...Array(n + 1).keys()].sort((a, b) => fv[a] - fv[b]);
    pts = ord.map((i) => pts[i]);
    fv = ord.map((i) => fv[i]);
    if (Math.abs(fv[n] - fv[0]) < tol) break;
    const c = Array(n).fill(0);                         // centroid of best n points
    for (let i = 0; i < n; i++) for (let j = 0; j < n; j++) c[j] += pts[i][j] / n;
    const xr = c.map((cj, j) => cj + (cj - pts[n][j]));  // reflect
    const fr = f(xr);
    if (fr < fv[0]) {
      const xe = c.map((cj, j) => cj + 2 * (cj - pts[n][j]));  // expand
      const fe = f(xe);
      if (fe < fr) { pts[n] = xe; fv[n] = fe; } else { pts[n] = xr; fv[n] = fr; }
    } else if (fr < fv[n - 1]) {
      pts[n] = xr; fv[n] = fr;
    } else {
      const xc = c.map((cj, j) => cj + 0.5 * (pts[n][j] - cj));  // contract
      const fc = f(xc);
      if (fc < fv[n]) { pts[n] = xc; fv[n] = fc; }
      else {                                            // shrink toward best
        for (let i = 1; i <= n; i++) {
          pts[i] = pts[i].map((pij, j) => pts[0][j] + 0.5 * (pij - pts[0][j]));
          fv[i] = f(pts[i]);
        }
      }
    }
  }
  return pts[0];
}

const f = (v) => (v[0] - 1) ** 2 + (v[1] - 2) ** 2;
console.log(nelderMead(f, [0, 0]));                     // -> [1, 2]
